src/net_processing.cpp

```diff
@@ -344,8 +346,10 @@ struct CNodeState {
          */
         std::multimap<std::chrono::microseconds, uint256> m_tx_process_time;
 
-        //! Store all the transactions a peer has recently announced
-        std::set<uint256> m_tx_announced;
+        /* Store all the transactions a peer has recently announced,
+         * along with their process time
+         */
+        std::map<uint256, std::chrono::microseconds> m_tx_announced;
```

Comment:
I think you'd need that to remain constant/log time in worst-case/attack scenarios because the announced hash is completely under the control of an attacker.